fix #926 (tutoriels en béta)

pierre-24 opened the pull request:
| Q | R |
| --- | --- |
| Correction de bugs ? | oui |
| Nouvelle Fonctionnalité ? | oui |
| Tickets concernés | #926 |

Comme @firm1 c'est occupé de s'assurer que les tutos étaient bien disponible en béta coté _back_ (voir #958), je me suis assuré que c'était également le cas coté _front_ et j'ai mis à jour les _templates_ correspondantes, du mieux que je peux.

J'ai également ajouté une option pour permettre la mise à jour de la béta directement (sans devoir désactiver puis réactiver le tout).

pierre-24 committed:
fix #926 (tutoriels en béta)

Comme @firm1 c'est occupé de s'assurer que les tutos étaient bien
disponible en béta coté back (voir #958), je me suis assuré que
c'était également le cas coté front et j'ai mis à jour les
templates correspondantes.

firm1 commented on templates/tutorial/base.html:
```diff
@@ -16,7 +16,11 @@
 {% block breadcrumb_base %}
-    <li><a href="{% url "zds.member.views.tutorials" %}">Mes tutoriels</a></li>
+    {% if user in tutorial.authors.all or perms.tutorial.change_tutorial %}
+        <li><a href="{% url "zds.member.views.tutorials" %}">Mes tutoriels</a></li>
```
Je ne sais pas si c'est une bonne idée de changer la racine des tutoriels si je suis sur mon tuto ou si je suis staff.

Parce que ça signifierai que tous les staff n'auront pas dans leur fil d'ariane "Tous les tutoriels", et que les auteurs auront un fil d'ariane incohérent.

AMHA l'ancienne version était bonne.

pierre-24 replied:
Ben j'y ai réfléchi, justement. L'ancienne version balance sur "mes tutoriels", ce qui est incohérent quand on est un user lambda qui voit un tuto en béta (selon moi). Après, c'est vrai que si on est staff, balancer sur "mes tutoriels" est pas la meilleure des idées.

Toi tu dit que tout devrait être "tout les tutoriels" ou j'ai pas compris ? Parce que l'ancienne version, c'était pas ça ^^

firm1 replied:
> Toi tu dit que tout devrait être "tout les tutoriels" ou j'ai pas compris

Tu as compris. En effet changer l'ancienne version

pierre-24 replied:
Donc selon toi quel devrait être le comportement : 
- Pour un user _lambda qui visite le tuto_
- Pour un auteur ?
- Pour un membre de staff ?

(ou alors modifier le test en `si bêta alors "lien tout-les-tutos" sinon "lien mes-tutos"` ?)

firm1 replied:
- Un user lambda, un auteur, ou un staff, quand ils visitent un tuto en version **online**, ils doivent avoir un truc du genre `Accueil > Tutoriels > Apprenez le C`
- Un auteur qui visite son tuto en version **offline**, doit avoir `Accueil > Mes Tutoriels > Apprenez le C`
- Un staff qui visite un tuto en version offline, doit avoir `Accueil > Tutoriels > Apprenez le C`
- Un auteur qui va sur l'onglet "Mes tutoriels" doit avoir `Accueil > Mes Tutoriels`

pierre-24 replied:
Oki, j'arrange ça ;)

firm1 commented on zds/tutorial/views.py:
```diff
@@ -433,6 +433,18 @@ def activ_beta(request, tutorial_pk, version):
 
 @can_write_and_read_now
 @login_required
+def update_beta(request, tutorial_pk, version):
```
Ce traitement doit être fait en POST pour éviter les CSRF

pierre-24 replied:
Je sais. Le truc c'est que je sais plus qui (Spacefox ?) c'était lancé dans l'action de passer tout en post (#497). Tu propose donc que je le fasse déjà pour ça ou je laisse ainsi en attendant Spacefox ?

firm1 replied:
Etant donné que tu rajoute une view, autant prendre déjà des requetes de type POST. Et donc, dans le template tu dois faire la correction, ça ne doit pas être un lien, mais un formulaire.

pierre-24 replied:
Je vais modifier ça :)

(c'est déjà un formulaire par contre, voir ligne 212 du template, juste qu'il balance en get, il suffit de mettre un champ _hidden_ et de passer en post)

firm1 commented on templates/tutorial/includes/tutorial_item.part.html:
```diff
@@ -1,5 +1,11 @@
-    <a href="{{ tutorial.get_absolute_url_online }}">
+    <a href="
+        {% if beta %}
+            {{ tutorial.get_absolute_url }}{% if tutorial.sha_beta %}?version={{tutorial.sha_beta}}{% endif %}
```
Pourquoi ne pas utiliser ici tout simplement [get_absolute_url_beta](https://github.com/pierre-24/zds-site/blob/fix_926/zds/tutorial/models.py#L124) plutot de se trainer dans variables dans le template ?

pierre-24 replied:
Parce que la moitié du temps, j'oublie qu'elle existe ^^

firm1 commented:
Bon je crois avoir finir de survoler cette PR, faudra qu'un QA passe dessus pour l'aspect fonctionnel.

Cependant, dans les templates, tu as rajouté pas mal de fois la même conditions `{% if user in tutorial.authors.all or perms.tutorial.change_tutorial %}`

Je pense qu'il aurait été préférable de rajouter pour chaque `view.html` des tutoriels, des templates `view-read-only.html`. Ainsi, chaque view.html étend la read-only conrrespondante. Et donc dans la view.html tu ne fais que rajouter les éléments qui rend ton template éditable.

pierre-24 commented:
C'est pas bète, mais je vois pas en quoi ça va réduire le nombre de `{% if user in ... %}`. Au mieux, ça facilite la lisibilité des templates (ce en quoi c'est tout à fait louable).

(et avant qu'on ne me le dise, je vais rajouter un test, promis :p )

pierre-24 committed:
Inclusion suggestion @firm1+test unitaire

- Activer/désactiver/updater la béta se fait maintenant en POST
- Modification du comportement du fil d'ariane
- Ajout d'un test unitaire: `test_workflow_beta_tuto` pour les
  big- et minituto.

pierre-24 committed:
Retrait des fonctions `*_beta()` et urls

sandhose commented:
Testé, et approuvé!
